apps/actions-dependencies-updater/src/caches.mts

```diff
@@ -0,0 +1,95 @@
+import * as log from "./logger.mjs";
+import { GithubShaToVersionCache } from "./github.mjs";
+import { UpdateTransaction } from "./updater.mjs";
+
+const CACHE_DIR = ".cache/";
+const SHA_TO_VERSION_CACHE = "sha-to-version.json";
+
+export function initialize(forceRefresh: boolean, optionalId?: string) {
+  const id = optionalId ? optionalId : Date.now();
+  const cacheExists = fs.existsSync(path.join(CACHE_DIR, SHA_TO_VERSION_CACHE));
+
+  if (cacheExists && forceRefresh) {
+    try {
+      // rename .cache to .cache-<timestamp>
+      const newCachePath = path.join(
+        CACHE_DIR,
+        SHA_TO_VERSION_CACHE.replace(".json", `-${id}.json`),
+      );
+      log.warn(
+        `Forcing cache refresh. Previous cache will be moved to ${newCachePath}`,
+      );
+      fs.renameSync(path.join(CACHE_DIR, SHA_TO_VERSION_CACHE), newCachePath);
+    } catch (e) {
+      log.error(`Failed to remove cache: ${e}`);
+    }
+  }
+
+  return {
+    shaToVersion: new Cache<GithubShaToVersionCache>(
+      true,
+      SHA_TO_VERSION_CACHE,
+    ),
+    updateTransactions: new Cache<UpdateTransaction>(
+      false,
+      `updates-${id}.json`,
+      {},
+    ),
+  };
+}
+
+class Cache<T> {
+  private filePath: string;
+  private cache: T;
+
+  constructor(initializeFromDisk: boolean, filePath: string, initial?: T) {
+    this.filePath = path.join(CACHE_DIR) + filePath;
+
+    if (initializeFromDisk) {
+      this.cache = this.load();
+    } else if (initial) {
+      this.cache = initial;
+    } else {
+      this.cache = {} as T;
```

Comment:
```suggestion
      this.cache = {}
```

We might not need this cast if we extend the generic

Reply:
Still need it.

```
{} is assignable to the constraint of type  T  , but T could be instantiated with a different subtype of constraint  Record<string, any>  .
```